### ISSUE TITLE
rabbitmq admin password that contains special characters causes the rabbitmq module to throw errors

### ISSUE THREAD

**Issue by TomRitserveldt** [voxpupuli/puppet-rabbitmq#754]
Using a rabbitmq admin password that contains special characters causes the rabbitmq module to throw errors, containing said admin password.

for example in the create part of:
`puppet-control-repo/modules/upstream/rabbitmq/lib/puppet/provider/rabbitmq_binding/rabbitmqadmin.rb`

> Execution of '/usr/local/bin/rabbitmqadmin declare binding --vhost=/ --user=admin --password=j(6$70ho>94%= -c /etc/rabbitmq/rabbitmqadmin.conf source=ALTERNATE_EXCHANGE destination=CATCH_ALL arguments={} routing_key= destination_type=queue' returned 1: *** Access refused: /api/bindings/%2F/e/ALTERNATE_EXCHANGE/q/CATCH_ALL

**Comment by wyardley** [comment]
What module version are you running? Thought this had been fixed (I know I’ve used special characters on passwords and created bindings in the past), but can take a look at this sometime if you can post your version of the module, of RabbitMQ, and of Puppet

**Comment by TomRitserveldt** [comment]
Our HEAD is currently at commit 8452f04c93c8cba52e8af20bc550f761ab953fae
from the 26th of October (2018).

we're waiting on this PR https://github.com/voxpupuli/puppet-rabbitmq/pull/739 to be merged to sync back to master on this repo.

rabbitmq version: rabbitmq-server 3.6.6-1
puppet server version: 5.3.4

**Comment by wyardley** [comment]
repro code:

```puppet
class { 'rabbitmq':
  service_manage    => true,
  port              => 5672,
  delete_guest_user => true,
  admin_enable      => true,
}

rabbitmq_user { 'admin':
  admin    => true,
  password => 'j(6$70ho>94%=',
  tags     => ['monitoring', 'tag1'],
} ->

rabbitmq_user_permissions { 'admin@/':
  configure_permission => '.*',
  read_permission      => '.*',
  write_permission     => '.*',
}   

rabbitmq_exchange { 'exchange2@/':
  user     => 'admin',
  password => 'j(6$70ho>94%=',
  type     => 'topic',
  ensure   => present,
} ->

rabbitmq_queue { 'queue2@/':
  user        => 'admin',
  password    => 'j(6$70ho>94%=',
  durable     => true,
  auto_delete => false,
  ensure      => present,
} ->

rabbitmq_binding { 'exchange2@queue2@/':
  user             => 'admin',
  password         => 'j(6$70ho>94%=',
  destination_type => 'queue',
  routing_key      => '#',
  ensure           => present,
}
```

**Comment by wyardley** [comment]
I know roughly how to fix this, but running into some oddities so far.

re: #739 I think the test(s) still need to be adjusted or updated, and I don't think the author has responded to any feedback there, so I can't say if / when it would be merged unless someone else wants to work on it more.

**Comment by wyardley** [comment]
The password does seem to get double-quoted by the shell exec. Some of the problem has to do with escaping, but I think there's also a separate issue in RabbitMQ itself with passwords containing a dollar sign (at least affecting the version I was testing with, and possibly others):

https://github.com/rabbitmq/rabbitmq-server/commit/381bf067b5be77f513e9fff18ffb9d32c910b505

```
[root@centos7-64-1 rabbitmq]# rabbitmqctl change_password admin 'j(6$70ho>94%='
Changing password for user "admin" ...
...done.
[root@centos7-64-1 rabbitmq]# /usr/local/bin/rabbitmqadmin declare binding --vhost=/ --user=admin --password='j(6$70ho>94%=' -c /etc/rabbitmq/rabbitmqadmin.conf source=exchange2 destination=queue2 arguments={} routing_key=# destination_type=queue
*** Access refused: /api/bindings/%2F/e/exchange2/q/queue2
[root@centos7-64-1 rabbitmq]# /usr/local/bin/rabbitmqadmin declare binding --vhost=/ --user=admin --password='j(6ho>94%=' -c /etc/rabbitmq/rabbitmqadmin.conf source=exchange2 destination=queue2 arguments={} routing_key=# destination_type=queue
*** Access refused: /api/bindings/%2F/e/exchange2/q/queue2
[root@centos7-64-1 rabbitmq]# /usr/local/bin/rabbitmqadmin declare binding --vhost=/ --user=admin --password='j(6\$70ho>94%=' -c /etc/rabbitmq/rabbitmqadmin.conf source=exchange2 destination=queue2 arguments={} routing_key=# destination_type=queue
*** Access refused: /api/bindings/%2F/e/exchange2/q/queue2
[root@centos7-64-1 rabbitmq]# rabbitmqctl change_password admin 'j(6\$70ho>94%='
Changing password for user "admin" ...
...done.
[root@centos7-64-1 rabbitmq]# /usr/local/bin/rabbitmqadmin declare binding --vhost=/ --user=admin --password='j(6$70ho>94%=' -c /etc/rabbitmq/rabbitmqadmin.conf source=exchange2 destination=queue2 arguments={} routing_key=# destination_type=queue
binding declared
```

So, escaping the dollar sign in the middle in the password for the user (which uses `rabbitmqctl`), but not for the others (which use `rabbitmqadmin`) seems to be a viable workaround.
e.g.,
```puppet
rabbitmq_user { 'admin':
  admin    => true,
  password => 'pa\$sword',
  tags     => ['monitoring', 'tag1'],
}

rabbitmq_exchange { 'exchange2@/':
  user     => 'admin',
  password => 'pa$sword',
  type     => 'topic',
  ensure   => present,
}
```
Thanks to @alexjfisher for some help bouncing ideas off of.

**Comment by TomRitserveldt** [comment]
I've been testing some variations and it seems like the dollar sign is the culprit when setting the (admin) user password. So currently a workaround for us would be to avoid the dollar sign in passwords.

As we are using rabbitmq 3.6.6, manually setting the password works fine:

> # rabbitmqctl change_password admin "j(6$70ho>94%="
But when you set the password (without escaping the dollar sign) in puppet, it seems to not properly set the password and thats when the other commands fail.

The problem persists on the latest version of this module(I pulled the latest from master on this repo).

**Comment by wyardley** [comment]
@TomRitserveldt if you're setting it that way, that's your problem. See the comment in https://github.com/voxpupuli/puppet-rabbitmq/issues/754#issuecomment-448864855  Try this instead:
```bash
# rabbitmqctl change_password admin 'j(6$70ho>94%='
```

To see what I mean, try:
```bash
$ echo "j(6$70ho>94%="
```

If you define both the password and the other resources in your Puppet config single-quoted with RabbitMQ 3.6.x and the newest version of the module, as in my example above, I think it should work properly.

Maybe test this self-contained example to be sure:
```puppet
rabbitmq_user { 'test':
  admin    => true,
  password => 'pa$sword',
  tags     => ['monitoring', 'tag1'],
}

rabbitmq_exchange { 'exchange2@/':
  user     => 'test',
  password => 'pa$sword',
  type     => 'topic',
  ensure   => present,
}
```

**Comment by wyardley** [comment]
I'm going to not flag this as a bug for now, since it seems that the behavior with older RMQ versions and the resulting inconsistency is a RMQ issue that has already been resolved. I'm going to close the issue too, but feel free to open if you come across any additional issues with quoting / escaping of passwords.

It may still be necessary to backslash escape certain special characters when using Puppet because the underlying libraries end up double-quoting when it shells out vs. single quoting. I haven't tested this super extensively, but I _think_ it shouldn't be necessary in most cases.
